quit/basicfunction.py

```diff
@@ -0,0 +1,121 @@
+import typing
+from collections.abc import Iterable
+
+import numpy as np
+from scipy import linalg
+
+
+def ket(psi: Iterable[complex]) -> np.ndarray:
+    """Create a ket |psi> as a vertical vector.
+
+    :param ket: Sequence of complex numbers.
+    :return: A vertical vector as numpy array.
+    """
+    return np.reshape(np.asarray(psi), (-1, 1))
+
+
+def basis(dim: int) -> Iterable[np.ndarray]:
+    """Create computational basis of dimension dim.
+
+    :param dim: Dimension of vector space.
+    :return: Set of {|i>}i where i={0,...,d-1}.
+    """
+    vec = [np.zeros(dim) for _ in range(dim)]
+    for i in range(dim):
+        vec[i][i] = 1
+    return vec
+
+
+def bra(psi: Iterable[complex]) -> np.ndarray:
+    """Create bra <psi| as a horizontal vector.
+
+    :param psi: Sequence of complex numbers.
+    :return: A horizontal vector as numpy array satisfying <psi| = |psi>^+
+    """
+    return np.transpose(np.conjugate(ket(psi)))
+
+
+def braket(psi: Iterable[complex], phi: Iterable[complex]) -> complex:
+    """Create braket as an inner product between vectors psi and phi.
+
+    :param psi: Sequence of complex numbers.
+    :param phi: Sequence of complex numbers.
+    :return: Inner product between vectors psi and phi.
+    """
+    return complex(bra(psi) @ ket(phi))
+
+
+def ketbra(psi: Iterable[complex], phi: Iterable[complex]) -> np.ndarray:
+    """Create a matrix from multiplication vectors psi and phi.
+
+    :param psi: Sequence of complex numbers.
+    :param phi: Sequence of complex numbers.
+    :return: A matrix |psi><phi|
+    """
+    return ket(psi) @ bra(phi)
+
+
+def proj(psi: Iterable[complex]) -> np.ndarray:
```

Comment:
The docstring should clearly state that the input `psi` does not need to be normalized, and is normalized by  `proj` instead.